src/blr/cli/clusterrmdup.py

```diff
@@ -12,237 +12,297 @@
+
+BUFFER_SIZE = 500
 
 
 def main(args):
     logger.info("Starting Analysis")
-
-    current_cache_rp = dict()
-    cache_reads = dict()
+    summary = Counter()
+    positions = OrderedDict()
     chrom_prev = None
-    pos_prev = None
+    pos_prev = 0
     merge_dict = dict()
-    cache_dup_pos = dict()
-    with pysam.AlignmentFile(args.input, "rb") as openin:
-        for read in tqdm(openin.fetch(until_eof=True), desc="Processing reads"):
-            summary["Total reads"] += 1
-
-            # Wait for mate of read until process
-            if read.query_name not in cache_reads:
-                cache_reads[read.query_name] = read
-                continue
-            else:
-                mate = cache_reads[read.query_name]
-                del cache_reads[read.query_name]
+    buffer_dup_pos = deque()
+    for read, mate in tqdm(parse_filtered_read_pairs(args.input, summary), desc="Reading pairs"):
+
+        # Get barcode and confirm that its not None
+        barcode = utils.get_bamtag(read, args.barcode_tag)
+        if not barcode:
+            summary["Non tagged reads"] += 2
+            continue
+
+        # Get orientation of read pair
+        if mate.is_read1 and read.is_read2:
+            orientation = "F"
+        else:
+            orientation = "R"
 
-            # Requirements: read mapped, mate mapped and read has barcode tag
-            rp_meet_requirements, bc_new = meet_requirements(read=read, mate=mate, barcode_tag=args.barcode_tag)
-            if rp_meet_requirements:
-                chrom_new = read.reference_name
-                summary["Reads analyced"] += 2
+        summary["Reads analyced"] += 2
 
-                pos_new = read.reference_start
-                rp_pos_tuple = (mate.reference_start, mate.reference_end, read.reference_start, read.reference_end)
+        chrom_new = read.reference_name
+        pos_new = read.reference_start
 
-                # Save all rp until position is new to know if any are marked as duplicates.
-                if chrom_new == chrom_prev and pos_new == pos_prev:
+        # Store position (5'-ends of R1 and R2) and orientation ('F' or 'R') with is used to group duplicates.
+        # Based on picard MarkDuplicates definition, see: https://sourceforge.net/p/samtools/mailman/message/25062576/
+        current_position = (mate.reference_start, read.reference_end, orientation)
 
-                    if rp_pos_tuple in current_cache_rp:
-                        current_cache_rp[rp_pos_tuple].add_read_pair_and_bc(read=read, mate=mate, bc=bc_new)
-                    else:
-                        current_cache_rp[rp_pos_tuple] = CacheReadPairTracker(rp_pos_tuple=rp_pos_tuple,
-                                                                              chromosome=chrom_new, read=read,
-                                                                              mate=mate, bc=bc_new)
+        if abs(pos_new - pos_prev) > BUFFER_SIZE or not chrom_new == chrom_prev:
```

Comment:
```suggestion
        if abs(pos_new - pos_prev) > BUFFER_SIZE or chrom_new  not == chrom_prev:
```

Reply:
Minor comment: Do we need such a big buffer size? It shouldn't be able to be more than a couple of bp right?